Sources/AsyncAlgorithms/Debounce/DebounceStorage.swift

```diff
@@ -0,0 +1,296 @@
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+final class DebounceStorage<Base: AsyncSequence, C: Clock>: @unchecked Sendable where Base: Sendable {
+    typealias Element = Base.Element
+
+    /// The state machine protected with a lock.
+    private let stateMachine: ManagedCriticalState<DebounceStateMachine<Base, C>>
+    /// The interval to debounce.
+    private let interval: C.Instant.Duration
+    /// The tolerance for the clock.
+    private let tolerance: C.Instant.Duration?
+    /// The clock.
+    private let clock: C
+
+    init(base: Base, interval: C.Instant.Duration, tolerance: C.Instant.Duration?, clock: C) {
+        self.stateMachine = .init(.init(base: base, clock: clock, interval: interval))
+        self.interval = interval
+        self.tolerance = tolerance
+        self.clock = clock
+    }
+
+    func iteratorDeinitialized() {
+        let action = self.stateMachine.withCriticalRegion { $0.iteratorDeinitialized() }
+
+        switch action {
+        case .cancelTaskAndUpstreamAndClockContinuations(
+            let task,
+            let upstreamContinuation,
+            let clockContinuation
+        ):
+            upstreamContinuation?.resume(throwing: CancellationError())
+            clockContinuation?.resume(throwing: CancellationError())
+
+            task.cancel()
+
+        case .none:
+            break
+        }
+    }
+
+    func next() async rethrows -> Element? {
+        // We need to handle cancellation here because we are creating a continuation
+        // and because we need to cancel the `Task` we created to consume the upstream
+        return try await withTaskCancellationHandler {
+            // We always suspend since we can never return an element right away
+
+            let result: Result<Element?, Error> = await withUnsafeContinuation { continuation in
+                self.stateMachine.withCriticalRegion {
+                    let action = $0.next(for: continuation)
+
+                    switch action {
+                    case .startTask(let base):
+                        self.startTask(
+                            stateMachine: &$0,
+                            base: base,
+                            downstreamContinuation: continuation
+                        )
+
+                    case .resumeUpstreamContinuation(let upstreamContinuation):
+                        // This is signalling the upstream task that is consuming the upstream
+                        // sequence to signal demand.
+                        upstreamContinuation?.resume(returning: ())
+
+                    case .resumeUpstreamAndClockContinuation(let upstreamContinuation, let clockContinuation, let deadline):
+                        // This is signalling the upstream task that is consuming the upstream
+                        // sequence to signal demand and start the clock task.
+                        upstreamContinuation?.resume(returning: ())
```

Comment:
I think perhaps we need to make sure to return these back out and resume them outside of the lock.

Reply:
See above comment but this ought to be totally safe